python/lsst/daf/butler/registry/_registry.py

```diff
@@ -820,12 +826,106 @@ def insertDatasets(self, datasetType: Union[DatasetType, str], dataIds: Iterable
                                              f"dimension row is missing.") from err
         return refs
 
-    def getDataset(self, id: int) -> Optional[DatasetRef]:
+    @transactional
+    def importDatasets(self, datasets: Iterable[DatasetRef],
+                       idGenerationMode: DatasetIdGenEnum = DatasetIdGenEnum.UNIQUE) -> List[DatasetRef]:
+        """Import one or more datasets into the `Registry`.
+
+        Difference from `insertDatasets` method is that this method accepts
+        `DatasetRef` instances which should already be resolved and have a
+        dataset ID. If registry supports globally-unique dataset IDs (e.g.
+        `uuid.UUID`) then datasets which already exist in the registry will be
+        ignored if imported again.
+
+        Parameters
+        ----------
+        datasets :  `~collections.abc.Iterable` of `DatasetRef`
+            Datasets to be inserted. All `DatasetRef` instances must have
+            identical ``datasetType`` and ``run`` attributes. ``run``
+            attribute can be `None` and defaults to ``self.defaults.run``.
+            Datasets can specify ``id`` attribute which will be used for
+            inserted datasets. All dataset IDs must have the same type
+            (`int` or `uuid.UUID`), if type of dataset IDs does not match
+            configured backend then IDs will be ignored and new IDs will be
+            generated by backend.
```

Comment:
I was considering this as a legit use case when for example importing from integer repo into UUID repo. If we disallow that then higher-level code will need to know how to convert ints into UUIDs (or other ID type) before import. Problem with that is that logic for calculating deterministic IDs is deliberately encapsulated into manager and high-level code does not need to care about that detail. If we want to allow high level code to do ID transform then we then need to expose datasets manager type to that code.

Reply:
That makes sense.  I think it might be another argument that `_importDatasets` should be a package-private method that's expected to be called by the import code, but probably not code outside daf_butler.

Reply:
I'm OK with making it semi-private though this probably will not stop determined people from using it. I'll rename it and also expand docstring to explain why it is private. I am still unsure what everyone's position on UUID generation is. Is what I implemented currently sufficient or do we really want to make clients responsible for generating UUIDs (both random and deterministic)?

Reply:
I suppose the question is, if I create a DatasetRef and attach a uuid myself, can I get that ref ingested as-is?

Reply:
The answer should be yes if you know that you are ingesting into UUID registry (which is what I want to try to hide from you).

Reply:
I'm actually happy for it to fail if a UUID DatasetRef is ingested into an integer registry.  UUID is meant to be the future. As time goes by there will be effectively zero old style registries. Once we've updated the main ones and change raw-ingest then if people want to ingest into an old registry can't they use the old software (or else have a configuration option in raw-ingest).

Reply:
I am happy with butler always being responsible for actually generating UUIDs.  I mostly just want to emphasize that the control we already give higher-level code in terms of what kind of UUID to generate still means there's a lot of things butler code can't do properly on its own (like know how to migrate existing repository content or correctly import integer-`dataset_id` exports).

Reply:
I agree that we still depend on client for consistently defining policy for deterministic IDs. If we want butler to be in complete control of that we can make that policy a part of registry configuration.

I'll change import to always raise if an import of UUIDs is attempted into integer registry.